packages/@lwc/template-compiler/src/codegen/helpers.ts

```diff
@@ -267,7 +266,11 @@ function collectStaticNodes(node: ChildNode, staticNodes: Set<ChildNode>, state:
         node.children.forEach((childNode) => {
             collectStaticNodes(childNode, staticNodes, state);
 
-            childrenAreStatic = childrenAreStatic && staticNodes.has(childNode);
+            childrenAreStatic &&= staticNodes.has(childNode);
+
+            // Bail out if any children have event listeners. Event listeners are only allowed at the top level of a
+            // static fragment, because the engine currently cannot attach listeners to nodes inside a static fragment.
+            childrenAreStatic &&= !isBaseElement(childNode) || childNode.listeners.length === 0;
```

Comment:
Here's an example of why this is necessary:

```html
<div>
  <button onClick={onClick}></button>
</div>
```

Without this code, we would hoist the `<div>` without any event listeners at all. Instead, we should consider the `<div>` non-static but the `<button>` static. I added some tests for this.